File: envs/el/ensemble_launcher/examples/python/mpi_example.py
import socket
import os
import psutil
import argparse
from time import sleep

import mpi4py.rc
mpi4py.rc.initialize = False
mpi4py.rc.finalize = False
from mpi4py import MPI

def parse_args() -> argparse.Namespace:
    """
    Parse command line arguments.
    Args:
        sleep_time: Time to sleep in seconds
        debug: Debug mode
    Returns:
        args: Parsed arguments
    """
    parser = argparse.ArgumentParser()
    parser.add_argument("--sleep_time", type=int, default=10, help="Time to sleep in seconds")
    parser.add_argument("--debug", action="store_true", help="Debug mode")
    return parser.parse_args()

def main() -> None:
    """
    A simple MPI application which sleeps for a given time.
    In debug mode, it will print information about the node, CPU and GPU each rank is bound to.
    """
    # Initialize MPI
    if not MPI.Is_initialized():
        MPI.Init()
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()
    rankl = int(os.getenv("PALS_LOCAL_RANKID"))

    # Parse command line arguments
    args = parse_args()

    # Print debug information
    if args.debug:
        hostname = socket.gethostname()
        p = psutil.Process()
        cpu_affinity = p.cpu_affinity()
        gpu_affinity = os.getenv("ZE_AFFINITY_MASK") or os.getenv("CUDA_VISIBLE_DEVICES")
        print(f"Hello from rank {rank}/{size} on node {hostname}, CPU cores {cpu_affinity} and GPU {gpu_affinity}", flush=True)

    # Sleep for the given time
    comm.Barrier()
    sleep(args.sleep_time)
    comm.Barrier()

    MPI.Finalize()

if __name__ == "__main__":
    main()

File: envs/el/ensemble_launcher/examples/c++/mpi_example.cpp
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <cstdlib>
#include <sched.h>
#include <mpi.h>

// util-linux-2.13-pre7/schedutils/taskset.c
void get_cores(char *str)
{ 
  cpu_set_t mask;
  sched_getaffinity(0, sizeof(cpu_set_t), &mask);

  char *ptr = str;
  int i, j, entry_made = 0;
  for (i = 0; i < CPU_SETSIZE; i++) {
    if (CPU_ISSET(i, &mask)) {
      int run = 0;
      entry_made = 1;
      for (j = i + 1; j < CPU_SETSIZE; j++) {
        if (CPU_ISSET(j, &mask)) run++;
        else break;
      }
      if (!run)
        sprintf(ptr, "%d,", i);
      else if (run == 1) {
        sprintf(ptr, "%d,%d,", i, i + 1);
        i++;
      } else {
        sprintf(ptr, "%d-%d,", i, i + run);
        i += run;
      }
      while (*ptr != 0) ptr++;
    }
  }
  ptr -= entry_made;
  *ptr = 0;
}

int main(int argc, char *argv[])
{
  /*
  A simple MPI application which sleeps for a given time.
  In debug mode, it will print information about the node, CPU and GPU each rank is bound to.
  */
  // Initialize MPI
  int rank, size;
  MPI_Init(&argc, &argv);
  MPI_Comm world = MPI_COMM_WORLD;
  MPI_Comm_rank(world, &rank);
  MPI_Comm_size(world, &size);
  //int rankl = int(std::getenv("PALS_LOCAL_RANKID"));

  // Parse command line arguments
  if (argc > 3) {
    printf("Usages:\n");
    printf("%s\n", argv[0]);
    printf("%s <sleep_time>\n", argv[0]);
    printf("%s <sleep_time> <debug true/false>\n", argv[0]);
    MPI_Finalize();
    return 1;
  }
  int sleep_time;
  bool debug;
  if (argc == 1) {
    sleep_time = 10;
    debug = false;
  } else if (argc == 2) {
    sleep_time = atoi(argv[1]);
    debug = false;
  } else {
    sleep_time = atoi(argv[1]);
    debug = strcmp(argv[2], "true") == 0;
  }

  // Print debug information
  if(debug) {
    char hostname[16];
    gethostname(hostname, 16);
    char list_cores[7*CPU_SETSIZE];
    get_cores(list_cores);
    const char* gpu_affinity = std::getenv("ZE_AFFINITY_MASK") ? std::getenv("ZE_AFFINITY_MASK") : std::getenv("CUDA_VISIBLE_DEVICES");
    printf("Hello from rank %d/%d on node %s, CPU cores %s and GPU %s\n", 
            rank, size, hostname, list_cores, gpu_affinity);
    fflush(stdout);
  }

  // Sleep for the given time
  MPI_Barrier(MPI_COMM_WORLD);
  sleep(sleep_time);
  MPI_Barrier(MPI_COMM_WORLD);
  
  MPI_Finalize();
  return 0;
}
